nixos/ec2-metadata-fetcher: decompress EC2 user data if compressed

johnnywalker opened the pull request:
## Description of changes

Decompress EC2 user data when compressed with `gzip`.

Closes https://github.com/NixOS/nixpkgs/issues/290336.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

This has been tested with uncompressed user data and user data compressed by Terraform's `base64gzip()` function. We are actively using these changes locally.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

tomfitzhenry reviewed (dismissed):
Nice fix, well explained in https://github.com/NixOS/nixpkgs/issues/290336

A few thoughts:
* Is amazon-init the only consumer of `user-data`? If not, is there risk that existing users rely on passing compressed payloads via user-data, and won't be expecting it to arrive uncompressed?
* Could you squash the tidy up commit https://github.com/NixOS/nixpkgs/pull/290346/commits/e5d28bd7595de0d37578417880e6e711253c3c82 please?

johnnywalker committed:
nixos/ec2-metadata-fetcher: decompress gzip-compressed user data

Closes https://github.com/NixOS/nixpkgs/issues/290336

johnnywalker commented on nixos/modules/virtualisation/ec2-metadata-fetcher.sh:
```diff
@@ -60,7 +60,18 @@ get_imds() {
+  case $ftype in
+    gzip*)
+      echo "decompressing: $1"
+      temp=$(mktemp)
```
Thanks for catching this! I previously debated about renaming the downloaded file but wasn't sure of the community's preference. I'm happy to do either and have declared a local variable for now.

Also, thanks for sharing the bit about case statements!

johnnywalker commented:
Thanks for the review @tomfitzhenry !

>     * Is amazon-init the only consumer of `user-data`? If not, is there risk that existing users rely on passing compressed payloads via user-data, and won't be expecting it to arrive uncompressed?

Great question. I'll take a stab at constraining this to `amazon-init` to avoid any unexpected surprises to existing users.

arianvp commented:
I think a note in the release notes would suffice  as well.

Most tools handling user-data (like `cloud-init`)  just talk to the metadata sever directly

philiptaron committed:
nixos/ec2-metadata-fetcher: harden try_decompress

- Declare `ftype` as local to avoid leaking into caller scope
- Skip decompression attempt on empty files
- Clean up temp file on decompression failure

philiptaron committed:
nixos/ec2-image: add test for gzip user-data decompression

Verify that fetch-ec2-metadata correctly decompresses gzip-compressed
EC2 user data. Adds a test helper for reuse by subsequent compression
format tests.

philiptaron committed:
nixos/ec2-metadata-fetcher: add bzip2 user-data decompression

Refactor try_decompress to use a decompress_cmd variable, making it
straightforward to add new compression formats. Add bzip2 support
and a corresponding NixOS test.

philiptaron committed:
nixos/ec2-metadata-fetcher: add xz user-data decompression

Support decompressing XZ-compressed EC2 user data via xzcat.
Includes a corresponding NixOS test.

philiptaron committed:
nixos/ec2-metadata-fetcher: add zstd user-data decompression

Support decompressing Zstandard-compressed EC2 user data via zstdcat.
Includes a corresponding NixOS test.

philiptaron committed:
nixos/ec2-metadata-fetcher: add lzip user-data decompression

Support decompressing lzip-compressed EC2 user data via lzip -dc.
Includes a corresponding NixOS test.

philiptaron commented:
Hey @johnnywalker, thanks for this contribution! The core idea of decompressing user data at the fetch layer so all downstream consumers benefit transparently is exactly right. I decided to go a little further.

I've rebased your branch onto current master and pushed a few additional commits on top. Here's what changed:

1. I hardened the `try_decompress` function a bit. It now declares `ftype` as local, has an early return for empty files, and cleans up the temp file if decompression fails. 
2. Then I refactored it to use a `decompress_cmd` variable, which made it easy to extend: on top of your original `gzip` support, I've added `bzip2`, `xz`, `zstd`, and `lzip` decompression as well, following the set of "normal" compression algorithms in the nixpkgs codebase.
3. Each format has its own commit and a corresponding NixOS test in ec2-image.nix that exercises the full fetch-ec2-metadata pipeline with compressed user data.

I also tested the gzip path end-to-end on a real EC2 instance by building a custom AMI from this branch, launching it in us-west-2 with gzip-compressed user data, and verifying that the decompressed script executed correctly via amazon-init. All looked good.

Baring any objections, I'm going to merge this PR.

philiptaron commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 290346 --additional-package nixosTests.ec2-image`
Commit: `d09e847072fc2710de347009c8f8cac6d48a190e`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 test built:</summary>
  <ul>
    <li>nixosTests.ec2-image</li>
  </ul>
</details>

johnnywalker commented:
Thanks @philiptaron ! I like the support for additional compression types. Hopefully this will prevent some headaches for years to come!
